### PR TITLE
Fix JsonSerializerOptions.GetConverter recreating built-in converters

### LINKED ISSUE #65770: Repeatedly initialize the static variables s_defaultSimpleConverters and s_defaultFactoryConverters when calling the method GetConverter of JsonSerializerOptions
### Description

https://github.com/dotnet/runtime/blob/a1a653c7217706fcb820914d65126234bc2447a8/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs#L162

https://github.com/dotnet/runtime/blob/a1a653c7217706fcb820914d65126234bc2447a8/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs#L27

### Reproduction Steps

According to the referenced code

### Expected behavior

Variables s_defaultSimpleConverters and s_defaultFactoryConverters are initialized only once

### Actual behavior

Variables s_defaultSimpleConverters and s_defaultFactoryConverters are repeatedly initialized with each call

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #65770.

@layomia do you think is this the right approach? I changed `GetConverter` so that it now calls `InitializeForReflectionSerializer` instead of `RootBuiltInConverter` since it already has a flag guarding access. The only difference is that additionally roots the reflection-based `JsonTypeInfo` factory delegate. I suspect this is fine since `GetConverter` is already marked `RequiresUnreferencedCode` and might actually fix a broken corner case (see changes in test code).

@ericstj Do you think this meets the bar for 6.0 servicing? I can imagine certain use cases where this could incur a substantial amount of added allocations (and it's a regression from 5.0)

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs [hunk 1]
```diff
@@ -99,7 +99,6 @@ private void EnsurePushCapacity()
         public JsonConverter Initialize(Type type, JsonSerializerOptions options, bool supportContinuation)
         {
             JsonTypeInfo jsonTypeInfo = options.GetOrAddJsonTypeInfoForRootType(type);
-            Debug.Assert(options == jsonTypeInfo.Options);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:102]: This assertion has been invalidated by #64646, however it wasn't being caught by our test suite until now.

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:102]: I assume it would still be true if the new comparer introduced in https://github.com/dotnet/runtime/pull/64646 were used.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs [hunk 2]
```diff
@@ -159,7 +159,10 @@ internal JsonConverter GetConverterFromMember(Type? parentClassType, Type proper
         [RequiresUnreferencedCode("Getting a converter for a type may require reflection which depends on unreferenced code.")]
         public JsonConverter GetConverter(Type typeToConvert!!)
         {
-            RootBuiltInConverters();
+            if (!IsInitializedForReflectionSerializer)
+            {
+                InitializeForReflectionSerializer();
+            }
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:165]: I don't think calling this method should root the reflection-based metadata gathering code paths, potentially masking future bugs with source-gen.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs:102]: `==` still checks for reference equality, the comparer is internal implementation detail when initializing the caches.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.cs [hunk 3]
```diff
@@ -161,19 +163,29 @@ public static void VerifyObjectConverterWithPreservedReferences()
...
-        [Fact]
+        [ConditionalFact(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
         public static void GetConverterRootsBuiltInConverters()
         {
-            JsonSerializerOptions options = new();
-            RunTest<DateTime>();
-            RunTest<Point_2D>();
-
-            void RunTest<TConverterReturn>()
+            RemoteExecutor.Invoke(static () =>
```
Comment (krwq) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.cs:169]: this feels a bit like a unit test rather than functional test because it tests an implementation detail. I'll suggest to move it to unit tests.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:165]: Thanks. What is the distinction between rooting the reflection-based metadata gathering paths and rooting the built-in converter factories that depend on said reflection-based metadata gathering paths? Under the current setup an options instance not configured for reflection will eagerly return converters that do require reflection.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:165]: Based on your remarks on this PR https://github.com/dotnet/runtime/pull/51897 I'm guessing that this is only meant as an entrypoint for reflection-based serialization anyways.

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:165]: > Under the current setup an options instance not configured for reflection will eagerly return converters that do require reflection.

Hmm what do you mean? The source generator doesn't use this method `GetConverter(Type)`, so we left the reflection to avoid breaking folks that used it pre .NET 6.0.
